### PR TITLE
ENH: added `random.gamma`

### PR DESCRIPTION
**Pull request by samir-nasibli**
## Description
gamma(shape[, scale, size]) Draw samples from a Gamma distribution.

## TODO:
* array_like of floats for `scale` and `shape` params

## Checklist

- [x] Docstrings for all functions
- [ ] Gallery example in ./doc/examples (new features only)
- [X] Unit tests:
- [X] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

### CODE REVIEW

**Comment by samir-nasibli** [comment]
Need to be redesigned, because supported devices are Host and CPU.

https://software.intel.com/content/www/us/en/develop/documentation/oneapi-mkl-dpcpp-developer-reference/top/random-number-generators/distributions/oneapi-mkl-rng-gamma.html

File: dpnp/random/_random.pyx [hunk 1]
```diff
@@ -111,6 +113,33 @@ cpdef dparray dpnp_exponential(double beta, size):
     return result
 
 
+cpdef dparray dpnp_gamma(double shape, double scale, size):
+    """
+    Return a random matrix with data from the "gamma" distribution.
```
Comment (oleksandr-pavlyk) [inline dpnp/random/_random.pyx:118]: Why is it call "random matrix" ? The code populates a flat buffer with random numbers. 

"Random matrix" has special meaning in computation math, and this has nothing to do with it.

Returns an array populated with samples from gamma distribution. 

Quotes around gamma are not needed.

File: tests/skipped_tests.tbl [hunk 2]
```diff
@@ -20,6 +20,8 @@ tests/test_linalg.py::test_eig_arange[300-float32]
 tests/test_linalg.py::test_eig_arange[300-int64]
 tests/test_linalg.py::test_eig_arange[300-int32]
 tests/test_random.py::test_randn_normal_distribution
+tests/test_random.py::test_radnom_gamma_seed
```
Comment (oleksandr-pavlyk) [inline tests/skipped_tests.tbl:23]: typo: should be ``test_random_gamma_seed``

File: dpnp/random/_random.pyx (hunk 1 above)
Comment (samir-nasibli) [reply dpnp/random/_random.pyx:118]: Thank you. This comment is common for all other functions. I will refactor on other PR

**Comment by oleksandr-pavlyk** [comment]
Something is amiss. Basic moments do not match:

```
In [24]: np.var(np.random.gamma(shape=2.56, scale=0.9, size=10**6))
Out[24]: 2.0689620288819444

In [25]: np.mean(np.random.gamma(shape=2.56, scale=0.8, size=10**6))
Out[25]: 2.0468521330146694

In [26]: np.var(np.random.gamma(shape=2.56, scale=0.8, size=10**6))
Out[26]: 1.6367525135977068

In [27]: dpnp.mean(dpnp.random.gamma(shape=2.56, scale=0.8, size=10**6))
Out[27]: 2.559302218900502

In [28]: dpnp.var(dpnp.random.gamma(shape=2.56, scale=0.8, size=10**6))
Out[28]: 6.545711234562073
```

File: dpnp/random/_random.pyx (hunk 1 above)
Comment (samir-nasibli) [reply dpnp/random/_random.pyx]: Fix for all rest functions on #186

**Comment by samir-nasibli** [comment]
> Something is amiss. Basic moments do not match:
> 
> ```
> In [24]: np.var(np.random.gamma(shape=2.56, scale=0.9, size=10**6))
> Out[24]: 2.0689620288819444
> 
> In [25]: np.mean(np.random.gamma(shape=2.56, scale=0.8, size=10**6))
> Out[25]: 2.0468521330146694
> 
> In [26]: np.var(np.random.gamma(shape=2.56, scale=0.8, size=10**6))
> Out[26]: 1.6367525135977068
> 
> In [27]: dpnp.mean(dpnp.random.gamma(shape=2.56, scale=0.8, size=10**6))
> Out[27]: 2.559302218900502
> 
> In [28]: dpnp.var(dpnp.random.gamma(shape=2.56, scale=0.8, size=10**6))
> Out[28]: 6.545711234562073
> ```

@oleksandr-pavlyk Thank you for the report. I have fixed it.
Reproduce:
````python
>>> np.var(np.random.gamma(shape=2.56, scale=0.8, size=10**6))
1.6414887360289188
>>> np.mean(np.random.gamma(shape=2.56, scale=0.8, size=10**6))
2.0480613902279794
>>> np.var(dpnp.random.gamma(shape=2.56, scale=0.8, size=10**6))
1.6337041157687837
>>> np.mean(dpnp.random.gamma(shape=2.56, scale=0.8, size=10**6))
2.049404203568917
```

**Review by shssf — APPROVED** [review]

**Comment by oleksandr-pavlyk** [comment]
Expectations of bounded functions, or moments for light-tailed distributions is appropriate.

An appropriate check is that sample expectation is within several numbers (say 5) of standard deviations of the estimator from its mean. The mean needs to be known analytically, while the standard deviation may be estimated from the sample

**Comment by oleksandr-pavlyk** [comment]
Having built `dpnp` with oneAPI 2020.8.0.1005:

```
In [1]: import dpnp
Running on: Intel(R) Gen9
DPCtrl SYCL queue used
SYCL kernels link time: 0.182214 (sec.)


In [2]: import dpctl

In [3]: dpctl.set_default_queue('opencl', 'cpu', 0)

In [5]: X = dpnp.random.gamma(shape=125.4, scale=0.02, size=10**5)
terminate called after throwing an instance of 'oneapi::mkl::unsupported_device'
  what():  oneMKL: rng/generate: Intel(R) Gen9 is not supported
Aborted (core dumped) 
``` 

I was expecting `dpnp` to use queue from dpctl which was just set to be CPU.

It would greatly enhance user's  experience if these C++ exceptions where intercepted and rethrown as Python exceptions.

**Comment by oleksandr-pavlyk** [comment]
> Having built `dpnp` with oneAPI 2020.8.0.1005 (beta 10):
> 
> ```
> In [1]: import dpnp
> Running on: Intel(R) Gen9
> DPCtrl SYCL queue used
> SYCL kernels link time: 0.182214 (sec.)
> 
> 
> In [2]: import dpctl
> 
> In [3]: dpctl.set_default_queue('opencl', 'cpu', 0)
> 
> In [5]: X = dpnp.random.gamma(shape=125.4, scale=0.02, size=10**5)
> terminate called after throwing an instance of 'oneapi::mkl::unsupported_device'
>   what():  oneMKL: rng/generate: Intel(R) Gen9 is not supported
> Aborted (core dumped) 
> ```
> 
> I was expecting `dpnp` to use queue from dpctl which was just set to be CPU.
> 

Random number generation is not honoring `dpctl` supplied queue because it is using the cached random engine, which does have a queue retrieved at the module initialization time stored in the instance, so while `DPNP_CURRENT_QUEUE` gives the queue associated with OpenCL CPU device, the engine still uses the Level-0 GPU device. 

@shssf I think this is a problem.

**Comment by samir-nasibli** [comment]
> It would greatly enhance user's experience if these C++ exceptions where intercepted and rethrown as Python exceptions.

 Added issue #201 for this. Will be fixed in other PR

**Comment by oleksandr-pavlyk** [comment]
`shape=0` parameter causes oneMKL exception. These can be either disallowed, or handled as a special case using `dpnp.zeros`.

**Comment by samir-nasibli** [comment]
> `shape=0` parameter causes oneMKL exception. These can be either disallowed, or handled as a special case using `dpnp.zeros`.

AttributeError: type object 'dpnp.dparray.dparray' has no attribute 'zeros'.
But I have fixed case, when `shape=0`

**Comment by oleksandr-pavlyk** [comment]
> > `shape=0` parameter causes oneMKL exception. These can be either disallowed, or handled as a special case using `dpnp.zeros`.
> 
> AttributeError: type object 'dpnp.dparray' has no attribute 'zeros'.
> But I have fixed case, when `shape=0` via `dparray.fill`

Fill works. `zeros` is not a method of `dparray`, it is function in module, `dpnp.zeros`. It will produce a new array, so instead of 

```
result = dpnp.array(size, dtype=dtype)
result.fill(0.0)
```

you could simply write

```
result = dpnp.zeros(size, dtype=dtype)
```

**Review by oleksandr-pavlyk — APPROVED** [review]
Further changes must be done outside of this PR.

**Review by shssf — APPROVED** [review]

**Pull request merged**